src/main/java/org/folio/helper/CheckinReceivePiecesHelper.java

```diff
@@ -526,192 +461,171 @@ protected Future<PoLineAndTitleById> getPoLineAndTitleById(List<String> poLineId
+          long successQty = results.result().list().stream()
+            .filter(Objects::isNull)
+            .count();
+          logger.debug("{} out of {} holdings successfully processed", successQty, results.result().size());
+        }
+      })
+      .mapEmpty();
   }
 
-  /**
-   * Returns "Fully Received" status if quantity of expected piece records is
-   * zero, otherwise checks how many received pieces. If quantity of received
-   * piece records is zero, returns "Awaiting Receipt" status, otherwise -
-   * "Partially Received"
-   *
-   * @param byPoLine
-   *          all piece records obtained by PO line id
-   * @param poLine
-   *          PO Line record representation from storage
-   * @return completable future holding calculated PO Line's receipt status
-   */
-  private Future<ReceiptStatus> calculatePoLineReceiptStatus(List<Piece> byPoLine, PoLine poLine, List<Piece> pieces) {
-    long expectedPiecesQuantity = byPoLine.stream()
-      .filter(piece -> EXPECTED_STATUSES.contains(piece.getReceivingStatus()))
-      .count();
-    // Fully Received: If receiving and there is no expected piece remaining
-    if (!poLine.getCheckinItems().equals(Boolean.TRUE) && expectedPiecesQuantity == 0) {
-      return Future.succeededFuture(FULLY_RECEIVED);
-    }
-    // Partially Received: In case there is at least one successfully received piece
-    if (StreamEx.of(pieces).anyMatch(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus()))) {
-      return Future.succeededFuture(PARTIALLY_RECEIVED);
+  private Future<Boolean> createHoldingsForChangedLocations(Piece piece, String instanceId, RequestContext requestContext) {
+    String holdingKey = buildProcessedHoldingKey(piece, instanceId);
+    if (!ifHoldingNotProcessed(holdingKey) || isRevertToOnOrder(piece)) {
+      return Future.succeededFuture(true);
     }
-    // Pieces were rolled-back to Expected. In this case we have to check if
-    // there is any Received piece in the storage
-    long receivedQty = byPoLine.stream()
-      .filter(piece -> RECEIVED_STATUSES.contains(piece.getReceivingStatus()))
-      .count();
-    return Future.succeededFuture(receivedQty == 0 ? AWAITING_RECEIPT : PARTIALLY_RECEIVED);
-  }
 
-  private Future<List<Piece>> getPiecesByPoLine(String poLineId, RequestContext requestContext) {
-    String query = String.format("poLineId==%s", poLineId);
-    return pieceStorageService.getPieces(Integer.MAX_VALUE, 0, query, requestContext)
-      .map(PieceCollection::getPieces);
+    String locationId = getLocationId(piece);
+    String holdingId = getHoldingId(piece);
+    Location location = new Location().withLocationId(locationId).withHoldingId(holdingId);
+    var locationContext = RequestContextUtil.createContextWithNewTenantId(requestContext, piece.getReceivingTenantId());
+    return inventoryHoldingManager.getOrCreateHoldingsRecord(instanceId, location, locationContext)
+      .compose(createdHoldingId -> {
```

Comment:
what if holding already existed, in this case previous name holdingId was more suitable

Reply:
`holdingId` is already used. Maybe `newHoldingId`. If holding already existed createdHoldingId will be same as holdingId.